### ISSUE TITLE
Inconsistent Marker size with `scaling=True` and anisotropic scaling

### ISSUE THREAD

**Issue by brisvag** [vispy/vispy#2357]
In napari/napari#2213 we noticed an odd behaviour with `Markers`, where their size was varying depending on the camera view direction. I boiled it down to this example:

```py
from vispy import scene
import numpy as np

# basic canvas and camera
canvas = scene.SceneCanvas(show=True)
view = canvas.central_widget.add_view()
view.camera = scene.cameras.ArcballCamera(fov=0)

# create points and apply scaling transform
points = scene.visuals.Markers(
    pos=np.array([[0, 0, 0]]),
    size=0.1,
    edge_width=0,
    antialias=0,
    scaling=True,
    parent=view.scene
)
points.transform = scene.transforms.STTransform(scale=(5, 1, 1))
```

If you rotate the camera, the point will change in size. If you remove the scaling transform, or you set `scaling=False` (if you do, also change the size to `20` or the point will be too small), the issue will be fixed. This pinpoints this passage as the root of the problem (guess who's to `blame`...):

https://github.com/vispy/vispy/blob/02929ef4d82ecc154d4e19ad63edfffd19af4f5e/vispy/visuals/markers.py#L48-L55

I'm not immediately sure where the problem lies. Maybe we should perform these size calculations in a different coordinate space, before/after the scaling is applied? @kevinyamauchi @almarklein @djhoese any suggestions?

**Comment by djhoese** [comment]
If you comment out your `.transform` in your example or make all the scale values the same `(5, 5, 5)` then it works fine I think.

**Comment by djhoese** [comment]
Ah that's what you meant. Sorry.

**Comment by brisvag** [comment]
I think I understand the issue. Scaling affects the transforms, but *not* `size`. A breakdown of the code:

```glsl
// move horizontally in framebuffer space
// then go to visual coordinates
vec4 x = $framebuffer_to_visual(fb_pos + vec4(big_float, 0, 0, 0));
// subtract position, so we get the visual-coordinate vector describing
// the "horizontal direction parallel to the screen" in screen space
x = (x - pos);
// multiply that direction by the `size` (in visual space) and add it to the pos
// this gives us a vector to the edge of the point, which we convert in screen space
vec4 size_vec = $visual_to_framebuffer(pos + normalize(x) * a_size);
// divide by `w` for perspective, and subtract pos
// this gives us the actual screen-space size of the point
$v_size = size_vec.x / size_vec.w - fb_pos.x / fb_pos.w;
```

I'm pretty sure the issue here is that `size` should be also affected by the transform, but it isnt. I think the solution involves getting the scaling out of the transform and into the shader, and to take the component along `x`.

Or there's maybe a way to do it with what we have already...

**Comment by djhoese** [comment]
So you're saying that `a_size` needs to make the round trip from visual space to FB space to visual space to see how it is effected by possible transformations that might include scaling? No, that can't be right. :thinking: 

One question: why is this logic only dealing with the `x` component?

We see in your example that if the scaling is the same for all components/axes that it works fine. Doesn't this suggest that the other components have to be included in the above logic?

**Comment by brisvag** [comment]
> One question: why is this logic only dealing with the `x` component?

Points are by definition isotropic in screen space, so any direction works as long as it's parallel to the screen. Basically, a single edge point is sufficient to determine their `gl_PointSize`, which is all we need.

However, we need to scale this size that comes out depending on the angle between this direction and the scaling. Something like this:

![image](https://user-images.githubusercontent.com/23482191/179928226-e7a26928-77eb-48b8-88da-9d845625cd7d.png)

Hopefully it's clear enough:
- `P` is the plane parallel to the screen (as seen by the camera `C`, which passes through the point.
- `d` is the direction vector we get along the screen space `x`, which could be pointing anywhere in *visual space*.
- the visual's `Vx` and `Vy` (horizontal) dimensions have different scaling (respectively `1` and `5`)

In the current shader, if our point size is `1` and `d` is perfectly aligned to `Vx`, we get `1`. However, if we're aligned to `Vy`, we get `5` (or *maybe* `1/5`, not 100% sure).

This is because we take a step of size `a_size` *in visual space*, which then gets scaled like everything else. We need this step *not* to be scaled, because while point *position* should be scaled, their size should not.

Any in-between angle will result in a combination (and don't forget there's a third dimension...).

Maybe it's just a matter of multiplying (or dividing) the direction vector by the "scale vector"?

**Comment by brisvag** [comment]
Closed by #2359.
